### ISSUE TITLE
Regent: replicate scalar tasks in a control replicated context

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#822]
From https://github.com/StanfordLegion/legion/issues/633#issuecomment-612585630, this is a reminder to myself to make sure scalar tasks (like the ones that Regent lifts from future operations) are replicated in a DCR context, to avoid broadcasting the future result of these tasks.

@lightsighter What's required to do this? I need to mark the tasks themselves as idempotent, and then what?

**Comment by lightsighter** [comment]
I think we might have reached the tipping point where it's time to have the runtime support "future computations", e.g. inline functions that run whenever a set of futures are ready and produce a new future value. They will look like very light-weight tasks that don't support launching any sub-computations or touching any regions. They will run on whatever processor they were launched on and will be automatically control replicated by the runtime because they by definition have to be pure functions of the futures they take as arguments. I think both Regent and Legate could benefit from them. If we don't want to support arbitrary computations, I think all of the use cases we have so far could be expressed as reductions on future values.

**Comment by elliottslaughter** [comment]
Well, in theory Regent can do this with any sort of expression on a future, so even things like array access and field access. I'm not sure those are expressed naturally as reductions.

Are these "future computations" going to match the task interface? I realize the existing task interface is wider than we really need, but that would avoid needing to rewrite a bunch of Regent task generation logic.

**Comment by magnatelee** [comment]
I'm also a big proponent of this runtime support as I'm currently using many tasks on futures and future maps to create index spaces and fields from futures. As a natural extension to this, I'd also want to describe computations on future maps that are mapped individually on the futures; many of my use cases transform one future map to another. I believe these computations are merely an index launch-lifted computations on futures.

I'd expect the task interface to be different as these future tasks don't need take any regions, but it'd be fine to keep the interface the same and pass an empty vector of `PhysicalRegion` objects. We may need to change the launcher interface though if we want to allow passing multiple future maps to a task.

**Comment by lightsighter** [comment]
> Are these "future computations" going to match the task interface?

No, they are going to have their own interface because I want to use the type system to make it clear that users aren't going to be able to do anything like launching sub-tasks or anything else in these computations. In C++ it will look like:

```
template<typename T>
T* (*fnptr)(const std::vector<Future> &futures)
```

in the C API it will look like

```
void preamble(const legion_future_t **futures, unsigned *num_futures);
void postamble(const void *retval, size_t retsize);
```

> As a natural extension to this, I'd also want to describe computations on future maps that are mapped individually on the futures

I think we're going to leave this for a different issue, as I was describing to you earlier that this brings up a bunch of questions of how to shard this for control replication and what to do if the sharding of two future maps does not align. I don't want to get in to that here. With computations on individual futures its obvious that we should just replicate this computation in all shards and we know that every shard will have its own copy of the futures and produce its own new futures.

> We may need to change the launcher interface though if we want to allow passing multiple future maps to a task.
This is already supported for tasks today, convert your future maps to argument maps (nearly zero-cost) and put them here:
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion.h#L1550-1552

**Comment by lightsighter** [comment]
Ok, so I ended up waffling on this one. I started implementing future functions and they were just looking too much like tasks so I instead went in another direction which is to allow you to annotate an individual task launch as a "local function" task (if you want to propose a different name let me know):

https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion.h#L1486-1493
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion/legion_c.h#L2770-2775

A "local function" task is a lighter-weight task that can only have future arguments (no regions supported), should be idempotent, and is not eligible for distribution by the mapper (the mapper will still be able to map it, but it can't pick a processor different than the local node). In control replicated contexts, the runtime will automatically replicate this task for you so that you get one copy per shard, which is safe because of the restrictions required by the "local function" annotation.

This is available in the master and control replication branches so go ahead and test it out.
